Categorical.(get|from)_dummies

clbarnes opened the pull request:
Simply converting categorical variables to and from dummy variables.

- [x] closes #8745
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Intentionally smaller-scoped than #31795 (and indeed `get_dummies`) as a broadly useful MVP which can be chained with other basic functionality. The tests are fairly rudimentary and I welcome any edge cases which should be picked out.

## For discussion

### from_dummies

* class method rather than free function: Keeps categorical-related functionality together, reduces surface in the global namespace, more obvious what is produced.
* silently drop column with NA header: wasn't sure about this. Maybe it should raise a warning?
* No handling of masked dataframes or dataframes with NA values
* No subsetting or renaming of columns: callers can do this themselves

### to_dummies

* Name: I went for `Categorical.to_dummies` instead of matching the free function `get_dummies`. The symmetry of to/from aids understanding, and using `get_` might imply A) that something is being got, which it isn't or B) signature/ feature parity with the existing method, which wasn't a design goal for me.
* `to_dummies` return type: `cls.to_dummies` returns bools, where `get_dummies` returns uint8s by default, which doesn't make a lot of sense to me as we are representing boolean data (and they're the same in memory anyway). Dummy variables are generally used for regression where a continuous variable is required, so ints don't get us any closer to what we may want, and being able to index into the categories using the row may be valuable.
* No `dtype` argument: I didn't see any benefit of `cls.to_dummies(dtype=float)` over `cls.to_dummies().astype(float)`. The latter is more explicit, no slower, and minimises API surface.
* No `prefix`, `prefix_sep` arguments: these unnecessarily assume string column headers. If someone wants to rename their columns, they can: IMO it's not a core requirement of this method.
* `dumma_na` replaced with `na_column`: if we're including the argument, we may as well let the user decide what they want to call their column, using the dtype they prefer (e.g. `"other"`, `-1` etc). They can always supply `np.nan` if they want `get_dummies`-like behaviour.
* No `sparse` argument: This I regret. Producing a sparse array would be valuable. However, it would drastically complicate the method, so I left it out for the MVP. The caller can always sparsify it after it's produced, so long as RAM isn't an issue for the temporary df.
* No `drop_first` argument: If someone wants to drop one of their columns, they can (`cls.to_dummies().drop(columns="my_col")`): again, not a core requirement of this method. Broadly speaking, I'm not in favour of adding arguments to save the caller <=1 line of their own code unless there are e.g. speed gains.

clbarnes committed:
Categorical (to|from)_dummies methods

Simplistic implementation to go between dummy variables and
Categoricals.

clbarnes committed:
Tests: Categorical.(to|from)_dummies

clbarnes committed:
Add reference to Categorical.to_dummies to get_dummies

MarcoGorelli commented on pandas/core/arrays/categorical.py:
```diff
@@ -379,6 +382,110 @@ def __init__(
+        Categories (3, object): [a, b, c]
+        """
+        # GH 8745
+        from pandas import Series
```
Is this only used to get `index_into`, which is only used to get `len(index_into)`? If so, why not just have
```python
mult_by = np.arange(1, df.shape[1]+1)
```?
?

```

clbarnes replied:
My bad, that was a hangover from before I started using `from_codes`.

MarcoGorelli commented on pandas/core/arrays/categorical.py:
```diff
@@ -379,6 +382,110 @@ def __init__(
         self._codes = coerce_indexer_dtype(codes, dtype.categories)
 
+    @classmethod
+    def from_dummies(cls, dummies: "DataFrame", ordered=None):
```
Can you annotate `ordered`, as well as the return type?

clbarnes replied:
Is there a preference for how to annotate the self return type? As I understand it, you can do it with a `TypeVar`, with a string, or by using `from __future__ import annotations` (although only in 3.7+).

MarcoGorelli commented on pandas/core/arrays/categorical.py:
```diff
@@ -379,6 +382,110 @@ def __init__(
+    @classmethod
+    def from_dummies(cls, dummies: "DataFrame", ordered=None):
+        """
+        Create a `Categorical` using a ``DataFrame`` encoding those categories
+        as dummy/ one-hot encoded variables.
```
The [docstring guide](https://pandas.pydata.org/docs/development/contributing_docstring.html) asks that the short summary fit on a single line, is there a way to reword it?

MarcoGorelli commented on pandas/core/arrays/categorical.py:
```diff
@@ -379,6 +382,110 @@ def __init__(
+        index_into = Series([np.nan] + list(df.columns))
+        mult_by = np.arange(1, len(index_into))
+
+        codes = (df.astype(int) * mult_by).sum(axis=1) - 1
+        codes[codes.isna()] = -1
```
I may have misunderstood what's happening here, but would it work to use `pd.factorize`?

clbarnes replied:
I may be missing what `factorize` does, but I don't think it would save any effort here -  factorize wants a 1D array (which this isn't until `codes` is instantiated), and then doesn't produce a `Categorical` unless given one, so the last line wouldn't change either.

MarcoGorelli replied:
OK yes, I was thinking something like
```python
codes, _ = factorize(df.idxmax(axis=1))
```
, but then it'd be necessary to deal with the case when there's a row without any ones, which'd require more code and would be no shorter/clearer than what you've already got. So, ignore the `factorize` suggestion :)

Could of thiings then:
- is `astype(int)` necessary? It should work without it
- perhaps there would be a comment explaining why you do `-1` at the end of this line
```python
        codes = (df.astype(int) * mult_by).sum(axis=1) - 1
```
?
Now that I've run the code, it's obvious what it does, but it wasn't when I was just reading it

clbarnes replied:
You're right, the `astype` wasn't necessary. There's a comment now, best way I could think of explaining it but I can go for prose if you prefer.

MarcoGorelli commented on pandas/tests/arrays/categorical/test_api.py:
```diff
@@ -399,6 +399,19 @@ def test_remove_unused_categories(self):
+            [1, 1.5, "a", (1, "b"), np.nan],
+        ],
+    )
+    def test_to_dummies(self, vals):
```
I think the issue number as a comment should go in every new test

MarcoGorelli commented on pandas/core/arrays/categorical.py:
```diff
@@ -379,6 +382,110 @@ def __init__(
+
+        Parameters
+        ----------
+            dummies : DataFrame of bool-like
+            ordered : bool
+                Whether or not this Categorical is ordered.
```
Move these back one indent

MarcoGorelli commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -255,6 +255,7 @@ Other enhancements
 - :meth:`HDFStore.put` now accepts `track_times` parameter. Parameter is passed to ``create_table`` method of ``PyTables`` (:issue:`32682`).
 - Make :class:`pandas.core.window.Rolling` and :class:`pandas.core.window.Expanding` iterable（:issue:`11704`)
 - Make ``option_context`` a :class:`contextlib.ContextDecorator`, which allows it to be used as a decorator over an entire function (:issue:`34253`).
+- :class:`~pandas.core.arrays.categorical.Categorical` now has methods for converting to and from dummy/ one-hot encoded variables: :meth:`Categorical.to_dummies` and :meth:`Categorical.from_dummies` respectively. :meth:`Categorical.to_dummies` is smaller in scope than :func:`~pandas.core.reshape.reshape.get_dummies`, which can still be used if you require the extra flexibility.
```
Put the issue number at the end

MarcoGorelli reviewed (changes requested):
Cool, thanks @clbarnes !

clbarnes committed:
whatsnew: add issue number to Categorical.(to|from)_dummies

clbarnes committed:
Review comments for dummies tests

clbarnes committed:
Review comments for dummies implementation

MarcoGorelli commented on pandas/core/arrays/categorical.py:
```diff
@@ -379,6 +382,107 @@ def __init__(
+        [a, b, c]
+        Categories (3, object): [a, b, c]
+        """
+        # GH 8745
```
I don't think the issue number needs to be here, should only be necessary in the tests

jreback reviewed (changes requested):
will look soon

jreback commented:
likely would want a doc-example of both of these near in https://pandas.pydata.org/docs/dev/user_guide/reshaping.html#computing-indicator-dummy-variables and/or links to categorical.rst

clbarnes committed:
dummies review comments

clbarnes committed:
User guide: Describe Categorical.(to|from)_dummies

clbarnes committed:
Fix user guide errors

clbarnes committed:
Fix numpy element from sequence error

clbarnes committed:
Test to_dummies column type cast

TomAugspurger commented on pandas/core/arrays/categorical.py:
```diff
@@ -379,6 +382,109 @@ def __init__(
+        --------
+        :func:`pandas.get_dummies`
+        """
+        from pandas import DataFrame, CategoricalIndex, Series
```
Why is this not using the existing `get_dummies` implementation? This shold be identical to `pd.get_dummies(Categorical)`, right?

TomAugspurger commented on pandas/core/arrays/categorical.py:
```diff
@@ -379,6 +382,109 @@ def __init__(
+        codes[codes.isna()] = -1
+        return cls.from_codes(codes, df.columns.values, ordered=ordered)
+
+    def to_dummies(self, na_column=None) -> "DataFrame":
```
I think this should be called `get_dummies` I don't think the justification for deviating from that name is strong enough to warrant a different name.

I also think it should exactly match the signature of `get_dummies`.

TomAugspurger commented on pandas/core/arrays/categorical.py:
```diff
@@ -379,6 +382,109 @@ def __init__(
+        The ``DataFrame`` must be coercible to boolean,
+        and have no more than one truthy value per row.
+        The columns of the ``DataFrame`` become the categories of the `Categorical`.
+        A column whose header is NA will be dropped;
+        any row with a NA value will be uncategorised.
```
Need examples showing these.

TomAugspurger commented on pandas/core/arrays/categorical.py:
```diff
@@ -379,6 +382,109 @@ def __init__(
+
+        Parameters
+        ----------
+        dummies : DataFrame of bool-like
```
Just the type on the first line. The structure can go on the second line.

TomAugspurger commented on pandas/core/arrays/categorical.py:
```diff
@@ -379,6 +382,109 @@ def __init__(
+        Raises
+        ------
+        ValueError
+            If a sample belongs to >1 category
```
Add an example.

TomAugspurger commented on pandas/tests/arrays/categorical/test_constructors.py:
```diff
@@ -643,3 +645,55 @@ def test_constructor_string_and_tuples(self):
+    def test_from_dummies_gt1(self):
+        # GH 8745
+        dummies = DataFrame([[1, 0, 1], [0, 1, 0], [0, 0, 1]], columns=["a", "b", "c"])
+        with pytest.raises(ValueError):
```
Make sure we're raising an informative error message here.

TomAugspurger commented on pandas/tests/arrays/categorical/test_constructors.py:
```diff
@@ -643,3 +645,55 @@ def test_constructor_string_and_tuples(self):
+
+    def test_from_dummies_nan(self):
+        # GH 8745
+        raw = ["a", "a", "b", "c", "c", "a", np.nan]
```
Is `nan` the only supported NA value that's dropped?

TomAugspurger commented on pandas/tests/arrays/categorical/test_constructors.py:
```diff
@@ -643,3 +645,55 @@ def test_constructor_string_and_tuples(self):
         expected_index = pd.Index([("a", "b"), ("b", "a"), "c"])
         assert c.categories.equals(expected_index)
+
+    def test_from_dummies(self):
```
Can you parametrize this over sparse = True/False?

clbarnes replied:
For the sake of simplicity I have not made an attempt to support loading from a sparse array but I can give it a go!

clbarnes reviewed (commented):
Thanks for your comments @TomAugspurger ! I've addressed most of them but think the `to_dummies` and `get_dummies` shared functionality is probably worth discussion on the main thread.

clbarnes committed:
Test review comments

- Test Categories.from_dummies from sparse
- Test that all NA-valued headers are dropped
- Test for informative error message

clbarnes committed:
Review comments for implementation

- Handle more NA types
- More examples
- More informative error message

clbarnes commented:
> I think this should be called get_dummies I don't think the justification for deviating from that name is strong enough to warrant a different name.
>
> I also think it should exactly match the signature of get_dummies.

@TomAugspurger My initial implementation was a thin wrapper around `get_dummies` but given how simple the "raw" implementation is I didn't see a huge benefit to it. I could switch back, although I had wondered whether it might be better for for `get_dummies` to actually call `to_dummies` in some cases to avoid some logic.

I do feel more strongly about keeping the name and API different to `get_dummies`, though. `get_` methods imply retrieving something from the object rather than converting it to a different form. `from_` is IMO the most obvious name for the alternate constructor, and if `from_` exists, the opposite ought to be `to_` (`as_` or `into_` would also be OK but I feel they imply more mutation/ destructive operations). I personally have a minor nit with the naming of pandas' IO methods on this basis - `read_csv` and `to_csv` are from two different opposite pairs, which increased the time it took for me to remember them.

My justifications for not supporting all of the `get_dummies` interface are in the PR description. Supporting the whole `get_dummies` API in `to_dummies` would necessitate supporting all of the reverse operations in `from_dummies` (which is the key part of this PR; `to_dummies` is more of an afterthought), which I considered to be a large complexity overhead for not a lot of gain in utility. If we're not supporting the whole API, they shouldn't have the same name.

clbarnes committed:
Fix doctest for missing values

clbarnes committed:
xfail for Categorical from sparse

clbarnes committed:
Fix tests

clbarnes committed:
fix isna

clbarnes committed:
Explicit integer type cast

clbarnes committed:
Test categorical <-> dummies roundtrip

clbarnes committed:
more type casts

TomAugspurger commented:
> I do feel more strongly about keeping the name and API different to get_dummies

Yeah, your reasons for `to_dummies` all make sense. But in my mind it comes down to a values judgement between

1. Symmetry the two methods, which favors `to_dummies` over `get_dummies`, as `to_dummies` aligns better with `from_dummies`
2. Consistency with the established name `get_dummies`.

In my opinion, 2 wins out.

> Supporting the whole get_dummies API in to_dummies would necessitate supporting all of the reverse operations in from_dummies

That would indeed be helpful. Which parameters / behaviors are especially hard to support? I'm happy to raise a `NotImplementedError` for things that could be done in a followup PR if anyone is interested.

jreback commented on doc/source/user_guide/categorical.rst:
```diff
@@ -129,6 +129,48 @@ This conversion is likewise done column by column:
+Some operations, like regression and classification,
+encodes a single categorical variable as a column for each category,
+with each row having False in all but one column (True).
+These are called dummy variables, or one-hot encoding.
```
can u add a link to a wiki page (or scikit lean ok too)

jreback commented on doc/source/user_guide/categorical.rst:
```diff
@@ -129,6 +129,48 @@ This conversion is likewise done column by column:
+
+The :meth:`pandas.Categorical.from_dummies` class method accepts a dataframe
+whose dtypes are coercible to boolean, and an ``ordered`` argument
+for whether the resulting ``Categorical`` should be considered ordered
```
you need to have some good cross links to the current get_dummies section

otherwise this is very confusing

i would prefer that these are actually in the get_dummies with just a small note here

jreback commented on pandas/core/arrays/categorical.py:
```diff
@@ -379,6 +382,137 @@ def __init__(
         self._dtype = self._dtype.update_dtype(dtype)
         self._codes = coerce_indexer_dtype(codes, dtype.categories)
 
+    @classmethod
```
i agree with the others

either remove this or make it identical to get_dummies

clbarnes replied:
I think the discussion around this was more about the necessity/ implementation of `to_dummies` - there currently isn't an equivalent of `from_dummies` in pandas, which is what originally precipitated the issue. But note taken re. `to_dummies` if that was the intention.

jreback commented on pandas/core/arrays/categorical.py:
```diff
@@ -379,6 +382,137 @@ def __init__(
+        #  010            020    2    1
+        #  001 * 1,2,3 => 003 -> 3 -> 2 = correct codes
+        #  100            100    1    0
+        codes = ((df * mult_by).sum(axis=1, skipna=False) - 1).astype("Int64")
```
this is actually a memory heavy impl

clbarnes replied:
Sure, I guess the alternative would be something like

```python
codes = pd.Series(np.full(len(df), np.nan), dtype="Int64")

row_totals = df.sum(axis=1, skipna=False)
... # multicat_rows check

codes[row_totals == 0] = -1
row_idx, code = np.nonzero(row_totals > 0)
codes[row_idx] = code
```

dsaxton commented:
@clbarnes Is this still active? Can you fix merge conflicts if so?

clbarnes commented:
Sorry I haven't been able to get back to this, I'll see what I can do today.

clbarnes committed:
Add wiki link for dummy variables

clbarnes committed:
Remove deprecated numpy <v1.16 check

clbarnes committed:
isort fix

clbarnes committed:
undo changes to whatsnew v1.1.0

clbarnes committed:
whatsnew/v1.2.0: Categorical get/from dummies

clbarnes committed:
Update user_guide/categorical docs

clbarnes committed:
Reference Categorical.get_dummies in reshape.py

clbarnes committed:
Categorical->dummies more like get_dummies

clbarnes committed:
categorical tests

clbarnes committed:
isort pandas_web

clbarnes committed:
fix _get_dummies_1d import path

clbarnes committed:
categorical.test_api: to->get dummies

clbarnes committed:
isort pandas_web

clbarnes committed:
fix typos in categorical doctests

clbarnes commented:
I'm calling `pandas.get_dummies` in `Categorical.get_dummies`. I'd rather be calling `pandas.core.reshape.reshape._get_dummies_1d`, but that fails a lint, and I don't want to start renaming existing functions - any guidance on that?

clbarnes committed:
isort test_datetime

clbarnes committed:
use get_dummies instead of _get_dummies_1d

clbarnes committed:
Reference get_dummies/ from_dummies in reshaping docs

clbarnes committed:
use prefix in from_dummies

clbarnes committed:
document prefix handling in categorical.rst

clbarnes commented:
Still to do:

- [ ] Edge cases for np.nonzero with "boolean" dtype in Categorical.from_dummies
- [ ] More tests, generally
- [ ] Check cohesiveness of docs

pep8speaks commented:
Hello @clbarnes! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/core/arrays/categorical.py`](https://github.com/pandas-dev/pandas/blob/765f0016604d2cc2522d3bd3ee91b2a3529e9a61/pandas/core/arrays/categorical.py):

> [Line 410:89](https://github.com/pandas-dev/pandas/blob/765f0016604d2cc2522d3bd3ee91b2a3529e9a61/pandas/core/arrays/categorical.py#L410): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 88 characters)

clbarnes committed:
Lower-memory impl for Categorical.from_dummies

jreback commented on pandas/core/arrays/categorical.py:
```diff
@@ -370,6 +373,224 @@ def __init__(
+        1  0.0  1.0  0.0
+        2  0.0  0.0  1.0
+        """
+        # Would be better to use pandas.core.reshape.reshape._get_dummies_1d
```
don't leave comments like this. why *cant* you use _get_dummies_1d ?

clbarnes replied:
Importing a function with a leading underscore fails a lint. It's one of the home-rolled lints, so it can't be ignored with `# noqa`.

jreback commented on pandas/core/arrays/categorical.py:
```diff
@@ -370,6 +373,224 @@ def __init__(
+
+        return cls.from_codes(codes.fillna(-1), cats, ordered=ordered)
+
+    def get_dummies(
```
pls type these

jreback commented on pandas/core/arrays/categorical.py:
```diff
@@ -370,6 +373,224 @@ def __init__(
+        from pandas import Series
+
+        copied = False
+        to_drop = dummies.columns[isna(dummies.columns.values)]
```
you don't need .values

jreback commented on pandas/core/arrays/categorical.py:
```diff
@@ -370,6 +373,224 @@ def __init__(
+
+        df = dummies.astype("boolean")
+        if fillna is not None:
+            df = df.fillna(fillna, inplace=copied)
```
i would kill the copied flag as its not worth it and adding un-needed complexity.

clbarnes replied:
There was a previous comment about the memory footprint so I thought I'd prevent additional copies where possible, but I would also prefer the simplicity.

jreback commented on pandas/core/arrays/categorical.py:
```diff
@@ -370,6 +373,224 @@ def __init__(
+        if fillna is not None:
+            df = df.fillna(fillna, inplace=copied)
+
+        row_totals = df.sum(axis=1, skipna=False)
```
why skipna?

clbarnes replied:
If there is no explicit `fillna` policy given, and there are still NA values in the data, I'd prefer to raise an error rather than silently pass bunk data through. Therefore nans should not be skipped in this step so that they can be checked for in the next line.

jreback commented on pandas/core/arrays/categorical.py:
```diff
@@ -370,6 +373,224 @@ def __init__(
+
+        row_totals = df.sum(axis=1, skipna=False)
+        if row_totals.isna().any():
+            raise ValueError("Unhandled NA values in dummy array")
```
is this tested?

clbarnes replied:
There are some holes left in the tests, on the to do list

jreback commented on pandas/core/arrays/categorical.py:
```diff
@@ -370,6 +373,224 @@ def __init__(
+        multicat_rows = row_totals > 1
+        if multicat_rows.any():
+            raise ValueError(
+                "{} record(s) belongs to multiple categories: {}".format(
```
use f-strings, is this tested?

clbarnes committed:
remove comment about use of _get_dummies_1d

clbarnes committed:
type-annotate get/from_dummies

clbarnes committed:
split overlong line

clbarnes committed:
blacken

clbarnes committed:
use f-strings

clbarnes committed:
add some typing

clbarnes committed:
remove unnecessary .values

arw2019 commented:
@clbarnes how's this going? If you're interested in continuing can you merge master & resolve conflicts and address comments?

clbarnes commented:
Sure, will do - did the first 90% but things got busy before I could tackle the last 90%. I'll try to get to this next week.

arw2019 commented:
> Sure, will do - did the first 90% but things got busy before I could tackle the last 90%. I'll try to get to this next week.

Sounds good (& no rush ofc).  Ping us when you're ready for the next round of reviews

jreback commented:
@clbarnes still active? I think this was pretty close.

mroeschke commented:
Thanks for the PR, but it appears to have gotten stale. Going to close for now but let us know if you can merge master and update and we'd be happy to reopen.
